// program/ControlUnit.js
// Part of <https://miracle.systems/p/walkner-snf> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const Lock = require('./Lock');

module.exports = ControlUnit;

function ControlUnit(broker, modbus, program, tagPrefix, index)
{
  /**
   * @protected
   * @type {h5.pubsub.Sandbox}
   */
  this.broker = broker.sandbox();

  /**
   * @protected
   * @type {object}
   */
  this.modbus = modbus;

  /**
   * @protected
   * @type {object}
   */
  this.program = program;

  /**
   * @private
   * @type {number}
   */
  this.index = index;

  /**
   * @private
   * @type {string}
   */
  this.tagPrefix = index > 0 ? `${tagPrefix}.${index}` : tagPrefix;

  /**
   * @private
   * @type {object.<string, Lock>}
   */
  this.locks = {};
}

/**
 * @returns {number}
 */
ControlUnit.prototype.getIndex = function()
{
  return this.index;
};

/**
 * @protected
 * @param {string} name
 * @returns {string}
 */
ControlUnit.prototype.getTagName = function(name)
{
  if (name.charAt(0) === '.')
  {
    return this.tagPrefix + name;
  }
  else
  {
    return name;
  }
};

/**
 * @protected
 * @param {string} tagName
 * @returns {string}
 */
ControlUnit.prototype.getTagValue = function(tagName)
{
  return this.modbus.values[this.getTagName(tagName)];
};

/**
 * @protected
 * @param {string} tagName
 * @returns {string}
 */
ControlUnit.prototype.getLastTagChangeTime = function(tagName)
{
  const tag = this.modbus.tags[this.getTagName(tagName)];

  if (_.isUndefined(tag))
  {
    return 0;
  }

  return tag.lastChangeTime;
};

/**
 * @protected
 * @param {string} tagName
 * @param {*} value
 * @param {function(Error|null, boolean)} [done]
 */
ControlUnit.prototype.setTagValue = function(tagName, value, done)
{
  if (!_.isFunction(done))
  {
    done = function() {};
  }

  tagName = this.getTagName(tagName);

  const tag = this.modbus.tags[tagName];

  if (!_.isObject(tag))
  {
    return done(new Error('UNKNOWN_TAG'));
  }

  if (tag.getValue() === value)
  {
    return done(null, true);
  }

  tag.writeValue(value, done);
};

/**
 * @protected
 * @param {string} tagName
 * @param {*} value
 * @param {number} [waitTime]
 * @param {function(Error|null, boolean)} done
 */
ControlUnit.prototype.ackTagValue = function(tagName, value, waitTime, done)
{
  if (arguments.length === 3)
  {
    done = waitTime;
    waitTime = 2000;
  }

  const tag = this.modbus.tags[this.getTagName(tagName)];

  if (_.isUndefined(tag) || tag.getValue() === value)
  {
    return done(null, false);
  }

  let sub;
  let timer;

  sub = this.broker.subscribe('tagValueChanged.' + tag.name)
    .on('message', function(message)
    {
      if (message.newValue === value)
      {
        sub.cancel();

        clearTimeout(timer);

        done(null, false);
      }
    });

  timer = setTimeout(
    function()
    {
      sub.cancel();

      done(new Error('ACK_TIMEOUT'), false);
    },
    waitTime || 2000
  );
};

/**
 * @protected
 * @param {string|Array.<string>} tagNames
 * @param {string} cb
 */
ControlUnit.prototype.watch = function(tagNames, cb)
{
  cb = this.bindExec(cb);

  if (!_.isArray(tagNames))
  {
    tagNames = [tagNames];
  }

  _.forEach(tagNames, (tagName) =>
  {
    this.broker.subscribe(`tagValueChanged.${this.getTagName(tagName)}`, cb);
  });

  process.nextTick(cb);
};

/**
 * @protected
 * @param {string} cb
 * @returns {function}
 */
ControlUnit.prototype.bindExec = function(cb)
{
  return this.program.exec.bind(
    null,
    this[cb].bind(this),
    `${this.tagPrefix}~${cb}`
  );
};

/**
 * @protected
 * @param {string} id
 * @returns {Lock}
 */
ControlUnit.prototype.lock = function(id)
{
  if (_.isUndefined(this.locks[id]))
  {
    this.locks[id] = new Lock();
  }

  return this.locks[id];
};

/**
 * @protected
 */
ControlUnit.prototype.debug = function(...args)
{
  this.program.debug(...args, {controlUnit: this.tagPrefix});
};

/**
 * @protected
 */
ControlUnit.prototype.warn = function(...args)
{
  this.program.warn(...args, {controlUnit: this.tagPrefix});
};

/**
 * @protected
 */
ControlUnit.prototype.error = function(...args)
{
  this.program.error(...args, {controlUnit: this.tagPrefix});
};
