// program/ProgramManager.js
// Part of <https://miracle.systems/p/walkner-snf> licensed under <CC BY-NC-SA 4.0>

'use strict';

var util = require('util');
var step = require('h5.step');
var ControlUnit = require('./ControlUnit');

module.exports = ProgramManager;

function ProgramManager(app)
{
  ControlUnit.call(this, app.broker, app.modbus, app.program, 'program');

  this.Program = app.mongoose.model('SnfProgram');

  this.currentProgram = null;

  this.broker.subscribe('snf.programs.edited', this.onProgramEdited.bind(this));

  this.watch(
    [
      '.30s', '.hrs', '.tester',
      'testKind.*',
      '.mnh',
      '.lampCount',
      '.lampBulb',
      '.lightSensors',
      '.limitSwitch',
      '.hrsCount.pc',
      '.hrsInterval.pc',
      '.hrsDuration',
      '.timeToShine',
      '.duration.pc',
      '.current.min',
      '.current.max',
      '.contactors',
      'light.1.min.plc',
      'light.2.min.plc',
      'light.1.min.pc',
      'light.2.min.pc',
      '.bulbHolder'
    ],
    'manageProgram'
  );
}

util.inherits(ProgramManager, ControlUnit);

ProgramManager.prototype.getCurrentProgram = function()
{
  return this.currentProgram;
};

ProgramManager.prototype.is30sKind = function()
{
  return this.getTagValue('testKind.1') && !this.getTagValue('testKind.2');
};

ProgramManager.prototype.isHrsKind = function()
{
  return !this.getTagValue('testKind.1') && !this.getTagValue('testKind.2');
};

ProgramManager.prototype.onProgramEdited = function(message)
{
  const newProgram = message.model;
  const oldProgram = this.getTagValue(`program.${newProgram.kind}`);

  if (newProgram._id !== oldProgram._id)
  {
    return;
  }

  this.setTagValue(`program.${newProgram.kind}`, newProgram, () =>
  {
    if (!this.getTagValue('contactors.2'))
    {
      this.changeProgram(newProgram.kind);
    }
  });
};

ProgramManager.prototype.manageProgram = function()
{
  if (this.is30sKind())
  {
    this.changeProgram('30s');
  }
  else if (this.isHrsKind())
  {
    this.changeProgram('hrs');
  }
  else
  {
    this.changeProgram('tester');
  }
};

ProgramManager.prototype.changeProgram = function(which)
{
  const lock = this.lock('changeProgram');

  if (lock.isLocked())
  {
    lock.cb = this.manageProgram.bind(this);

    return;
  }

  lock.on();

  const program = this.getTagValue(`.${which}`);

  this.updateProgramTags(
    program && program._id ? new this.Program(program) : null,
    lock
  );
};

ProgramManager.prototype.updateProgramTags = function(program, lock)
{
  const controlUnit = this;
  const noProgram = !program;
  const contactorsBytes = noProgram ? 0 : program.getContactorsBytes();

  this.currentProgram = program || null;

  if (!program)
  {
    program = {};
  }

  step(
    function setTagsStep()
    {
      controlUnit.setTagValue('.mnh', program.lampCount === 3, this.parallel());
      controlUnit.setTagValue('.lampCount', program.lampCount === 2, this.parallel());
      controlUnit.setTagValue('.lampBulb', program.lampBulb, this.parallel());
      controlUnit.setTagValue('.lightSensors', !!program.lightSensors, this.parallel());
      controlUnit.setTagValue('.limitSwitch', program.limitSwitch === true, this.parallel());
      controlUnit.setTagValue('.hrsCount.pc', program.hrsCount || 0, this.parallel());
      controlUnit.setTagValue('.hrsInterval.pc', program.hrsInterval || 0, this.parallel());
      controlUnit.setTagValue('.hrsDuration', program.hrsTime || 0, this.parallel());
      controlUnit.setTagValue('.timeToShine', program.waitForStartTime || 0, this.parallel());
      controlUnit.setTagValue('.duration.pc', program.illuminationTime || 0, this.parallel());
      controlUnit.setTagValue('.current.min', program.minCurrent || 0, this.parallel());
      controlUnit.setTagValue('.current.max', program.maxCurrent || 0, this.parallel());
      controlUnit.setTagValue('.bulbHolder', program.bulbHolder === true, this.parallel());

      if (controlUnit.getTagValue('.contactors') !== contactorsBytes)
      {
        this.setContactors = true;

        controlUnit.setTagValue('.contactors', 0, this.parallel());
      }

      controlUnit.setTagValue(
        'light.1.min.plc', controlUnit.getTagValue('light.1.min.pc') || 0, this.parallel()
      );
      controlUnit.setTagValue(
        'light.2.min.plc', controlUnit.getTagValue('light.2.min.pc') || 0, this.parallel()
      );
    },
    function()
    {
      if (this.setContactors)
      {
        setTimeout(this.next(), 1000);
      }
    },
    function()
    {
      if (this.setContactors)
      {
        controlUnit.setTagValue('.contactors', contactorsBytes, this.next());
      }
    },
    function()
    {
      lock.off();
    }
  );
};
